### PR TITLE
Concatenate output images

### PR DESCRIPTION
**Pull request by cbachhuber**
- Show live output during execution at 2x magnification
- Optionally show and save all outputs in one image, see below. Enables more direct visual comparison. Enabled per default, only needs a bool switch to go back to previous behavior.

outputs_iter-6

I thought something like this would be helpful when I planned how to find out where the remaining inaccuracies are (#23). The first thing I see in the above image is that subjectively, the particles are spread very far around the vehicles. Intuitively, I would expect a much tighter fit of the particle clusters to the vehicles. Do you agree with this or is my expectation wrong?

A significantly lower persistence probability (~0.5) achieves a tighter fit of particle clusters to the ground truth vehicles. This of course is far away from Nuss' parameters. What do you think is the right way to go here?

P.S.: I noticed that when disabling the image creation/display/saving altogether, my DOGM cycle times are improved:

```
# Everything enabled, as on this branch
  Minimum: 71ms
  Median:  83ms
  Mean:    82.8571ms
  Maximum: 102ms
```

```
# On this branch, main.cpp:84 commented out
  Minimum: 66ms
  Median:  70ms
  Mean:    71.1429ms
  Maximum: 85ms
```

I don't know how to explain this :sweat_smile: Maybe something related to CUDA 'cools down' during image saving?

### CODE REVIEW

**Comment by TheCodez** [comment]
Cool idea, thanks! 

> Intuitively, I would expect a much tighter fit of the particle clusters to the vehicles. Do you agree with this or is my expectation wrong?

I agree, but to be honest I'm not sure if that's correct behaviour or not. My best bet would be that some improvements to the implementation could improve this. Finding such errors is however pretty hard (it took roughly 5 months to find all errors that prevented the filter from running correctly :smile:). And given that we have no real ground thruth it makes it even harder. https://github.com/mitkina/dogma has the same behaviour so maybe it's correct after all, or both implementation are wrong :laughing:.

> A significantly lower persistence probability (~0.5) achieves a tighter fit of particle clusters to the ground truth vehicles. This of course is far away from Nuss' parameters. What do you think is the right way to go here?

This does indeed improve the particle clouds but at the expense of a higher velocity error. So this is the wrong thing to do.

> P.S.: I noticed that when disabling the image creation/display/saving altogether, my DOGM cycle times are improved:

Interesting :grinning: It's always nice to see even lower execution times👍 
It might be related to using [unified memory](https://devblogs.nvidia.com/unified-memory-cuda-beginners/) to simplify the image generation.

File: dogm/demo/utils/image_creation.cpp [hunk 1]
```diff
@@ -178,3 +178,53 @@ cv::Mat compute_particles_image(const dogm::DOGM& grid_map)
...
+    cv::Mat dogm_img = compute_dogm_image(grid_map, cells_with_velocity);
+
+    cv::Mat image_to_show{};
+    if (concatenate_images)
+    {
+        cv::hconcat(dogm_img, particle_img, image_to_show);
+        cv::hconcat(image_to_show, raw_meas_grid_img, image_to_show);
+
+        cv::Mat subtitles(image_to_show.rows * 0.2f, image_to_show.cols, image_to_show.type(), cv::Scalar::all(30));
+        add_text("Grid", 0.13f, subtitles);
```
Comment (TheCodez) [inline dogm/demo/utils/image_creation.cpp:208]: If you apply the text to the images before concatenating them, then the text position would be the same for all images. Wouldn't that be simpler?

Comment (cbachhuber) [reply dogm/demo/utils/image_creation.cpp]: True, good idea! Implemented.

**Comment by cbachhuber** [comment]
Got it, thank you for the explanation. This actually lead me to check how particles are visualized. And currently, as soon as one particle is in a cell, it is set to red (255 in red channel). With 3*10^5 particles, many cells will have very many particles in them. So instead of directly setting the red channel to full color, I suggest to increment it in [this commit](https://github.com/TheCodez/dynamic-occupancy-grid-map/pull/45/commits/e2d88fea4689dd75995999e7eb7259711ccd6b44). That leads to this much more expected output image:

outputs_step_11

So before, outliers that don't really contribute were shown too strongly. I guess particles are working as expected :+1: 

5 months to find that error? Respect for powering through that phase! :smile:

**Comment by TheCodez** [comment]
Wow this looks awesome :open_mouth:
That's the second time a wrong visualization has thrown me off :smile:
In the end it was worth it I guess :smile:

**Comment by cbachhuber** [comment]
Lessons learned are probably that being able to see what is actually going on (-> visualization in this case) cannot be underestimated :smile:

File: dogm/demo/utils/image_creation.cpp [hunk 2]
```diff
@@ -170,11 +171,83 @@ cv::Mat compute_particles_image(const dogm::DOGM& grid_map)
...
+    {
+        cv::imwrite(cv::format("raw_grid_step_%d.png", step + 1), raw_meas_grid_img);
+        cv::imwrite(cv::format("particles_step_%d.png", step + 1), particle_img);
+        cv::imwrite(cv::format("dogm_step_%d.png", step + 1), dogm_img);
+        image_to_show = dogm_img;
+    }
+
+    if (show_during_execution)
+    {
+        cv::namedWindow("DOGM", CV_WINDOW_NORMAL);
```
Comment (TheCodez) [inline dogm/demo/utils/image_creation.cpp:248]: Can you use `cv::WINDOW_NORMAL` instead `CV_WINDOW_NORMAL` to fix the window build?
What OpenCV version are you using? I'm using 4.1.1. We should atleast use the same the major version and note that in the README.

Comment (cbachhuber) [reply dogm/demo/utils/image_creation.cpp:248]: I'm using what `apt install libopencv-dev` gives me: 3.2.0. Your suggestion also works with the ubuntu build on opencv 3.2.0, let's use that. 

Good idea to note the opencv version down in the readme :+1: 
Currently, it looks like this library is working with both versions of opencv. We could also  support both versions until we run into an unsolvable issue.

Comment (TheCodez) [reply dogm/demo/utils/image_creation.cpp]: Alright, we should require the users to have OpenCV 3 or higher installed.

**Comment by TheCodez** [comment]
> P.S.: I noticed that when disabling the image creation/display/saving altogether, my DOGM cycle times are improved:

This should be solved now.

**Comment by cbachhuber** [comment]
Thank you for this! I'm still seeing a difference, though:

Everything as on master:
```
DOGM cycle stats (14 splits):
  Minimum: 63ms
  Median:  70ms
  Mean:    69.6429ms
  Maximum: 85ms
```

No image creation:
```
DOGM cycle stats (14 splits):
  Minimum: 62ms
  Median:  63ms
  Mean:    64.3571ms
  Maximum: 72ms
```

No image creation an no precision evaluation:
```
DOGM cycle stats (14 splits):
  Minimum: 73ms
  Median:  82ms
  Mean:    81.9286ms
  Maximum: 94ms
```

Those are weird, do you see similar timings?

**Comment by TheCodez** [comment]
On master:
```
DOGM cycle stats (14 splits):
  Minimum: 81ms
  Median:  87ms
  Mean:    85.6429ms
  Maximum: 89ms
```

No image creation:
```
DOGM cycle stats (14 splits):
  Minimum: 80ms
  Median:  85ms
  Mean:    84.2857ms
  Maximum: 87ms
```

No image creation an no precision evaluation:
```
DOGM cycle stats (14 splits):
  Minimum: 76ms
  Median:  78ms
  Mean:    79.2857ms
  Maximum: 87ms
```

Maybe there's really some kind of CUDA slow down. The timing also really fluctates sometimes.

**Comment by cbachhuber** [comment]
Interesting. I guess we will get to the reasons for this in future, after we learned more about the code's behavior.

**Pull request merged**